fix: Blockaid PPOM infinite loading spinner on controller error & refactor: clean ppom-middleware

digiwand opened the pull request:
## **Description**

- Fixes issue where usePPOM error is not caught causing an infinite loading spinner by adding a catch statement
- Fixes error messaging: "Error validating JSON RPC using PPOM" was not accurately representing some errors
- Clean up + TS refinement
- Updates related tests

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/23480?quickstart=1)

## **Related issues**

Fixes: https://github.com/MetaMask/metamask-extension/issues/23571
Tests Related to: https://github.com/MetaMask/metamask-extension/pull/23745

## **Manual testing steps**

1. Go to test dapp
2. Test PPOM Malicious transactions

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### **Before**

<!-- [screenshots/recordings] -->

### **After**

<!-- [screenshots/recordings] -->

## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've clearly explained what problem this PR is solving and how it is solved.
- [ ] I've linked related issues
- [ ] I've included manual testing steps
- [ ] I've included screenshots/recordings if applicable
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.
- [ ] I’ve properly set the pull request status:
  - [ ] In case it's not yet "ready for review", I've set it to "draft".
  - [ ] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

digiwand committed:
refactor: add description to SecurityAlertResponse type

digiwand committed:
refactor: ppom-middleware cleanup
- remove redundant req.securityAlertResponse

digiwand committed:
Merge branch 'develop' into refactor-clean-ppom-middleware

jpuri reviewed (dismissed):
Nice cleanup 👍

MajorLift commented on app/scripts/lib/ppom/ppom-middleware.ts:
```diff
@@ -70,56 +70,52 @@ export function createPPOMMiddleware(
+            securityAlertResponse = await ppom.validateJsonRpc(req);
+            securityAlertResponse.securityAlertId = securityAlertId;
+          })
+          .catch((error: any) => {
+            const errorObject = error as unknown as Error;
```
How would you feel about a) typing `error` with `unknown` instead of `any`, and b) using a runtime type check rather than a compile time type assertion to express assumptions about the shape of `error` (it has a `name` and `message` property)?

This diff only affirms the constraints already in place and adds handling for the exception case without adding new restrictions:

```diff
diff --git a/app/scripts/lib/ppom/ppom-middleware.ts b/app/scripts/lib/ppom/ppom-middleware.ts
index 47a1be0a98..32fda30fdc 100644
--- a/app/scripts/lib/ppom/ppom-middleware.ts
+++ b/app/scripts/lib/ppom/ppom-middleware.ts
@@ -81,17 +95,21 @@ export function createPPOMMiddleware(
             securityAlertResponse = await ppom.validateJsonRpc(req);
             securityAlertResponse.securityAlertId = securityAlertId;
           })
-          .catch((error: any) => {
-            const errorObject = error as unknown as Error;
+          .catch((error: unknown) => {
+            if (error instanceof Error) {
+              const errorObject = error;
 
-            sentry?.captureException(error);
-            console.error('Error validating JSON RPC using PPOM: ', error);
+              sentry?.captureException(error);
+              console.error('Error validating JSON RPC using PPOM: ', error);
 
-            securityAlertResponse = {
-              result_type: BlockaidResultType.Errored,
-              reason: BlockaidReason.errored,
-              description: `${errorObject.name}: ${errorObject.message}`,
-            };
+              securityAlertResponse = {
+                result_type: BlockaidResultType.Errored,
+                reason: BlockaidReason.errored,
+                description: `${errorObject.name}: ${errorObject.message}`,
+              };
+            } else {
+              throw error;
+            }
           })
           .finally(() => {
             updateSecurityAlertResponseByTxId(req, {
```

digiwand replied:
I like these suggestions! Thanks @MajorLift  

one thing is, I think we'd still like to capture the error in Sentry no matter what form it arrives in. What if, in the case we receive a non-Error type, we coerce the value into a string to turn it into an Error?

example: 
```
.catch((error: unknown) => {
            const errorObject =
              error instanceof Error ? error : new Error(error as string);

            sentry?.captureException(errorObject);
            console.error(
              'Error validating JSON RPC using PPOM: ',
              errorObject,
            );

            securityAlertResponse = {
              result_type: BlockaidResultType.Errored,
              reason: BlockaidReason.errored,
              description: `${errorObject.name}: ${errorObject.message}`,
            };
          })
```

MajorLift replied:
In that case, I would suggest the following: 

```ts
          .catch((error: unknown) => {
            sentry?.captureException(error);
            console.error(
              'Error validating JSON RPC using PPOM: ',
              isObject(error) || typeof error === 'string'
                ? error
                : JSON.stringify(error),
            );
            
            securityAlertResponse = {
              result_type: BlockaidResultType.Errored,
              reason: BlockaidReason.errored,
              description:
                error instanceof Error
                  ? `${error.name}: ${error.message}`
                  : JSON.stringify(error),
            };
          })
```

- `new Error(error as string)` doesn't actually change `error` into a string at runtime ([TypeScript has type assertions not type casts](https://github.com/microsoft/TypeScript/wiki/FAQ#why-doesnt-this-incorrect-cast-throw-a-runtime-error)), which means information that was in `error` would be lost in the resulting `Error` object.
- `String(error)` would be the way to do runtime conversion, but the result is likely to be `[object Object]`, which is why we need `JSON.stringify` (except in `console.error`, which can print out objects).

digiwand replied:
re: `new Error(error as string)`
oops. of course. not sure why I wrote it like this. I was thinking `new Error(String(error)),`, but you have a great point with it might turn out to be `[object Object]` as well. Good call with the `JSON.stringify(error)`

thanks @MajorLift! lgtm and updated https://github.com/MetaMask/metamask-extension/pull/23480/commits/06530bc22544aa49242181ccfca9280b4b440498

MajorLift commented on app/scripts/lib/ppom/ppom-middleware.ts:
```diff
@@ -70,56 +70,52 @@ export function createPPOMMiddleware(
+
+        req.securityAlertResponse = { ...securityAlertResponse };
       }
     } catch (error: any) {
       const errorObject = error as unknown as Error;
+
```
Same as above https://github.com/MetaMask/metamask-extension/pull/23480/files#r1529064707

```diff
diff --git a/app/scripts/lib/ppom/ppom-middleware.ts b/app/scripts/lib/ppom/ppom-middleware.ts
index 47a1be0a98..32fda30fdc 100644
--- a/app/scripts/lib/ppom/ppom-middleware.ts
+++ b/app/scripts/lib/ppom/ppom-middleware.ts
@@ -110,17 +128,21 @@ export function createPPOMMiddleware(
         req.securityAlertResponse = { ...securityAlertResponse };
       }
-    } catch (error: any) {
-      const errorObject = error as unknown as Error;
+    } catch (error: unknown) {
+      if (error instanceof Error) {
+        const errorObject = error;
 
-      sentry?.captureException(error);
-      console.error('Error validating JSON RPC using PPOM: ', error);
+        sentry?.captureException(error);
+        console.error('Error validating JSON RPC using PPOM: ', error);
 
-      req.securityAlertResponse = {
-        result_type: BlockaidResultType.Errored,
-        reason: BlockaidReason.errored,
-        description: `${errorObject.name}: ${errorObject.message}`,
-      };
+        req.securityAlertResponse = {
+          result_type: BlockaidResultType.Errored,
+          reason: BlockaidReason.errored,
+          description: `${errorObject.name}: ${errorObject.message}`,
+        };
+      } else {
+        throw error;
+      }
     } finally {
       next();
     }
```

MajorLift replied:
Same as above https://github.com/MetaMask/metamask-extension/pull/23480#discussion_r1530559068

```ts
    } catch (error: unknown) {
      sentry?.captureException(error);
      console.error(
        'Error validating JSON RPC using PPOM: ',
        isObject(error) || typeof error === 'string'
          ? error
          : JSON.stringify(error),
      );

      req.securityAlertResponse = {
        result_type: BlockaidResultType.Errored,
        reason: BlockaidReason.errored,
        description:
          error instanceof Error
            ? `${error.name}: ${error.message}`
            : JSON.stringify(error),
      };
    } finally {
```

digiwand replied:
updated https://github.com/MetaMask/metamask-extension/commit/06530bc22544aa49242181ccfca9280b4b440498

MajorLift commented:
Some additional typing suggestions on this file that fall outside the scope of the current diffs. Feel free to apply or disregard!

```diff
diff --git a/app/scripts/lib/ppom/ppom-middleware.ts b/app/scripts/lib/ppom/ppom-middleware.ts
index 47a1be0a98..977c49cdcf 100644
--- a/app/scripts/lib/ppom/ppom-middleware.ts
+++ b/app/scripts/lib/ppom/ppom-middleware.ts
@@ -1,6 +1,8 @@
+import { JsonRpcRequest, JsonRpcResponse } from 'json-rpc-engine';
 import { PPOM } from '@blockaid/ppom_release';
 import { PPOMController } from '@metamask/ppom-validator';
 import { NetworkController } from '@metamask/network-controller';
+import { Hex, Json, JsonRpcParams } from '@metamask/utils';
 import { v4 as uuid } from 'uuid';
 
 import {
@@ -12,7 +14,7 @@ import { SIGNING_METHODS } from '../../../../shared/constants/transaction';
 import { PreferencesController } from '../../controllers/preferences';
 import { SecurityAlertResponse } from '../transaction/util';
 
-const { sentry } = global as any;
+const { sentry } = global;
 
 const CONFIRMATION_METHODS = Object.freeze([
   'eth_sendRawTransaction',
@@ -20,7 +22,7 @@ const CONFIRMATION_METHODS = Object.freeze([
   ...SIGNING_METHODS,
 ]);
 
-export const SUPPORTED_CHAIN_IDS: string[] = [
+export const SUPPORTED_CHAIN_IDS: Hex[] = [
   CHAIN_IDS.ARBITRUM,
   CHAIN_IDS.AVALANCHE,
   CHAIN_IDS.BASE,
@@ -48,17 +50,28 @@ export const SUPPORTED_CHAIN_IDS: string[] = [
  * @param updateSecurityAlertResponseByTxId
  * @returns PPOMMiddleware function.
  */
-export function createPPOMMiddleware(
+export function createPPOMMiddleware<
+  Params extends JsonRpcParams = JsonRpcParams,
+  Result extends Json = Json,
+>(
   ppomController: PPOMController,
   preferencesController: PreferencesController,
   networkController: NetworkController,
   appStateController: any,
   updateSecurityAlertResponseByTxId: (
-    req: any,
+    req: JsonRpcRequest<JsonRpcParams> & {
+      securityAlertResponse: SecurityAlertResponse;
+    },
     securityAlertResponse: SecurityAlertResponse,
   ) => void,
 ) {
-  return async (req: any, _res: any, next: () => void) => {
+  return async (
+    req: JsonRpcRequest<Params> & {
+      securityAlertResponse: SecurityAlertResponse;
+    },
+    _res: JsonRpcResponse<Result>,
+    next: () => void,
+  ) => {
     try {
       const securityAlertsEnabled =
         preferencesController.store.getState()?.securityAlertsEnabled;
```

digiwand committed:
Merge branch 'develop' into refactor-clean-ppom-middleware

digiwand committed:
refactor: ppom-middleware error handling w/ unknown type
- see https://github.com/MetaMask/metamask-extension/pull/23480#discussion_r1529064707

Co-authored-by: Jongsun Suh <34228073+MajorLift@users.noreply.github.com>

digiwand committed:
refactor: ppom-middleware update TS for req/res and more
- https://github.com/MetaMask/metamask-extension/pull/23480#issuecomment-2004733717

digiwand committed:
Merge branch 'develop' into refactor-clean-ppom-middleware

digiwand commented:
@MajorLift nice! updated https://github.com/MetaMask/metamask-extension/pull/23480#issuecomment-2004733717 (sorry missed finalizing co-author on this one but comment is linked)

MajorLift commented:
This should probably fix the test errors. No worries on the co-author thing. :)

```diff
diff --git a/app/scripts/lib/ppom/ppom-middleware.test.ts b/app/scripts/lib/ppom/ppom-middleware.test.ts
index de3cff83f9..227ad69b52 100644
--- a/app/scripts/lib/ppom/ppom-middleware.test.ts
+++ b/app/scripts/lib/ppom/ppom-middleware.test.ts
@@ -1,3 +1,4 @@
+import { JsonRpcRequestStruct, JsonRpcResponseStruct } from '@metamask/utils';
 import { CHAIN_IDS } from '../../../../shared/constants/network';
 import {
   BlockaidReason,
@@ -64,8 +65,8 @@ describe('PPOMMiddleware', () => {
     const usePPOMMock = jest.fn();
     const middlewareFunction = createMiddleWare(usePPOMMock);
     await middlewareFunction(
-      { method: 'eth_sendTransaction' },
-      undefined,
+      { ...JsonRpcRequestStruct, method: 'eth_sendTransaction' },
+      { ...JsonRpcResponseStruct },
       () => undefined,
     );
     expect(usePPOMMock).toHaveBeenCalledTimes(1);
@@ -75,10 +76,15 @@ describe('PPOMMiddleware', () => {
     const usePPOM = async () => Promise.resolve('VALIDATION_RESULT');
     const middlewareFunction = createMiddleWare(usePPOM);
     const req = {
+      ...JsonRpcRequestStruct,
       method: 'eth_sendTransaction',
       securityAlertResponse: undefined,
     };
-    await middlewareFunction(req, undefined, () => undefined);
+    await middlewareFunction(
+      req,
+      { ...JsonRpcResponseStruct },
+      () => undefined,
+    );
     expect(req.securityAlertResponse).toBeDefined();
   });
 
@@ -86,10 +92,15 @@ describe('PPOMMiddleware', () => {
     const usePPOM = async () => Promise.resolve('VALIDATION_RESULT');
     const middlewareFunction = createMiddleWare(usePPOM, false);
     const req = {
+      ...JsonRpcRequestStruct,
       method: 'eth_sendTransaction',
       securityAlertResponse: undefined,
     };
-    await middlewareFunction(req, undefined, () => undefined);
+    await middlewareFunction(
+      req,
+      { ...JsonRpcResponseStruct },
+      () => undefined,
+    );
     expect(req.securityAlertResponse).toBeUndefined();
   });
 
@@ -97,10 +108,15 @@ describe('PPOMMiddleware', () => {
     const usePPOM = async () => Promise.resolve('VALIDATION_RESULT');
     const middlewareFunction = createMiddleWare(usePPOM, false, '0x2');
     const req = {
+      ...JsonRpcRequestStruct,
       method: 'eth_sendTransaction',
       securityAlertResponse: undefined,
     };
-    await middlewareFunction(req, undefined, () => undefined);
+    await middlewareFunction(
+      req,
+      { ...JsonRpcResponseStruct },
+      () => undefined,
+    );
     expect(req.securityAlertResponse).toBeUndefined();
   });
 
@@ -110,10 +126,15 @@ describe('PPOMMiddleware', () => {
     };
     const middlewareFunction = createMiddleWare({ usePPOM });
     const req = {
+      ...JsonRpcRequestStruct,
       method: 'eth_sendTransaction',
       securityAlertResponse: undefined,
     };
-    await middlewareFunction(req, undefined, () => undefined);
+    await middlewareFunction(
+      req,
+      { ...JsonRpcResponseStruct },
+      () => undefined,
+    );
     expect((req.securityAlertResponse as any)?.result_type).toBe(
       BlockaidResultType.Errored,
     );
@@ -132,8 +153,8 @@ describe('PPOMMiddleware', () => {
     const middlewareFunction = createMiddleWare(usePPOM);
     const nextMock = jest.fn();
     await middlewareFunction(
-      { method: 'eth_sendTransaction' },
-      undefined,
+      { ...JsonRpcRequestStruct, method: 'eth_sendTransaction' },
+      { ...JsonRpcResponseStruct },
       nextMock,
     );
     expect(nextMock).toHaveBeenCalledTimes(1);
@@ -146,8 +167,8 @@ describe('PPOMMiddleware', () => {
     const middlewareFunction = createMiddleWare(usePPOM);
     const nextMock = jest.fn();
     await middlewareFunction(
-      { method: 'eth_sendTransaction' },
-      undefined,
+      { ...JsonRpcRequestStruct, method: 'eth_sendTransaction' },
+      { ...JsonRpcResponseStruct },
       nextMock,
     );
     expect(nextMock).toHaveBeenCalledTimes(1);
@@ -163,8 +184,8 @@ describe('PPOMMiddleware', () => {
     };
     const middlewareFunction = createMiddleWare(usePPOM);
     await middlewareFunction(
-      { method: 'eth_sendTransaction' },
-      undefined,
+      { ...JsonRpcRequestStruct, method: 'eth_sendTransaction' },
+      { ...JsonRpcResponseStruct },
       () => undefined,
     );
     expect(validateMock).toHaveBeenCalledTimes(1);
@@ -180,8 +201,8 @@ describe('PPOMMiddleware', () => {
     };
     const middlewareFunction = createMiddleWare(usePPOM);
     await middlewareFunction(
-      { method: 'eth_someRequest' },
-      undefined,
+      { ...JsonRpcRequestStruct, method: 'eth_someRequest' },
+      { ...JsonRpcResponseStruct },
       () => undefined,
     );
     expect(validateMock).toHaveBeenCalledTimes(0);
@@ -189,6 +210,7 @@ describe('PPOMMiddleware', () => {
 
   it('normalizes transaction requests before validation', async () => {
     const requestMock1 = {
+      ...JsonRpcRequestStruct,
       method: 'eth_sendTransaction',
       params: [{ data: '0x1' }],
     };
@@ -212,7 +234,11 @@ describe('PPOMMiddleware', () => {
 
     const middlewareFunction = createMiddleWare(usePPOM);
 
-    await middlewareFunction(requestMock1, undefined, () => undefined);
+    await middlewareFunction(
+      requestMock1,
+      { ...JsonRpcResponseStruct },
+      () => undefined,
+    );
 
     expect(normalizePPOMRequestMock).toHaveBeenCalledTimes(1);
     expect(normalizePPOMRequestMock).toHaveBeenCalledWith(requestMock1);
diff --git a/app/scripts/lib/transaction/util.ts b/app/scripts/lib/transaction/util.ts
index ad22e7a4f7..1a06a50a99 100644
--- a/app/scripts/lib/transaction/util.ts
+++ b/app/scripts/lib/transaction/util.ts
@@ -14,6 +14,7 @@ import {
 } from '@metamask/user-operation-controller';
 ///: BEGIN:ONLY_INCLUDE_IF(blockaid)
 import { PPOMController } from '@metamask/ppom-validator';
+import type { Hex } from '@metamask/utils';
 import { captureException } from '@sentry/browser';
 import { addHexPrefix } from 'ethereumjs-util';
 import { v4 as uuid } from 'uuid';
@@ -60,7 +61,7 @@ export type AddTransactionOptions = NonNullable<
 >;
 
 type BaseAddTransactionRequest = {
-  chainId: string;
+  chainId: Hex;
   networkClientId: string;
   ppomController: PPOMController;
   securityAlertsEnabled: boolean;
```

cryptotavares commented on app/scripts/lib/ppom/ppom-middleware.ts:
```diff
@@ -1,6 +1,8 @@
 import { PPOMController } from '@metamask/ppom-validator';
 import { NetworkController } from '@metamask/network-controller';
+import { Hex, Json, JsonRpcParams } from '@metamask/utils';
+import { JsonRpcRequest, JsonRpcResponse } from 'json-rpc-engine';
```
`JsonRpcRequest` and `JsonRpcResponse` should be imported from `@metamask/utils`. `json-rpc-engine` is actualy re-exporting those two types from `@metamask/utils` and the repo has been archived.

digiwand replied:
ty! updated f47ec1c32e2d02cfa775b6900ce586f32c95be12

cryptotavares commented on app/scripts/lib/ppom/ppom-middleware.ts:
```diff
@@ -71,65 +84,71 @@ export function createPPOMMiddleware(
+          })
+          .catch((error: unknown) => {
+            sentry?.captureException(error);
+            console.error(
+              'Error validating JSON RPC using PPOM: ',
+              typeof error === 'object' || typeof error === 'string'
+                ? error
+                : JSON.stringify(error),
+            );
 
-              return securityAlertResponse;
-            }
+            securityAlertResponse = {
+              result_type: BlockaidResultType.Errored,
+              reason: BlockaidReason.errored,
+              description:
+                error instanceof Error
+                  ? `${error.name}: ${error.message}`
+                  : JSON.stringify(error),
+            };
```
This is going to result in capturing errors that are not directly related with the ppom validation... `usePPOM` might throw an error that is completely unrelated with validating request. For example any of [these](https://github.com/MetaMask/ppom-validator/blob/main/src/ppom-controller.ts#L319-L329) can throw an unrelated error. And I might be wrong, but I am assuming that `'User has securityAlertsEnabled set to false'` should not be a blockaid error description and captured within the securityAlertResponse.

Nevertheless those errors are not being caught anywhere (usePPOM returns a promise and we are not awaiting for it in order to be able to catch it [here](https://github.com/MetaMask/metamask-extension/blob/develop/app/scripts/lib/ppom/ppom-middleware.ts#L125)). So I do agree that we need to either await for usePPOM to resolve or if we do not need to await for it then we need to add the chained catch. But we need to distinguish between errors being thrown within the validation callback and errors being thrown by usePPOM.

digiwand committed:
refactor: import JsonRpcRequest,
  JsonRpcResponse from @metamask/util

digiwand committed:
fix: TS lint errors ppom-middleware
- https://github.com/MetaMask/metamask-extension/pull/23480#issuecomment-2008925985

digiwand committed:
refactor: restory usePPOM try/catch and update error messages

cryptotavares commented on app/scripts/lib/ppom/ppom-middleware.ts:
```diff
@@ -71,65 +89,91 @@ export function createPPOMMiddleware(
-              const securityAlertResponse = {
+              console.error(
+                'Error validating JSON RPC using PPOM: ',
+                typeof error === 'object' || typeof error === 'string'
+                  ? error
+                  : JSON.stringify(error),
```
Does not need to be solved in this PR, but why are we doing this serialization?
I get that JSON.stringify(error) when error is an instanceof Error will result in an empty object (as Error properties are not enumerable). But why typeof 'object'?
Wouldn't this be a better approach?
```js
error instanceof Error || typeof error === 'string'
```
For example the above option would allow us to properly log objects with nested arrays of objects.

Also this pattern is repeated a couple of times in this file. Should we at least have a method to `consoleLogSerializedError`? It would just have two args: the error (unknown) and the message to be logged.

Anyways, serliazing an unknown variable properly can be complex, as I am sure I am missing a couple of edge cases.

MajorLift replied:
`console.error` outputs interactively expandable representations of input objects. This applies to nested objects/arrays as well, so to my understanding, serialization isn't necessary here unless we're dealing with very large, deeply nested objects. 

`typeof error === 'object'` is because I assumed that we'd want the `console.error` representation for thrown object literals in general, not just `Error` types.

Although I'm not familiar with the logging and monitoring pipeline in the extension, so my assumptions might be off here.

I think a `consoleLogSerializedError` helper method makes a lot of sense. We have [`serializeError`](https://github.com/MetaMask/rpc-errors/blob/aa0754d3574ea3394126b752240caf93a8a9ea10/src/utils.ts#L104) and related methods in `@metamask/rpc-errors` for json-rpc errors. Something similar for browser console and sentry errors would definitely be useful.

cryptotavares replied:
Oh you are 100% right.. I was thinking about node console and not the browser (which is the env where the extension is executed). My bad 🤦‍♂️ 
@digiwand feel free to ignore my comment about `typeof x === 'object'` vs `instanceof error`.

Yes we should have a common way to not just serialize errors, but also to stack them up (where the stack trace and message of one error does not get lost in the hierarchical error handling). Anyways this out of scope of this PR.

digiwand commented:
thanks for the reviews and discussions! Merging and fixing/updating tests in a new PR to get this fix in sooner

digiwand commented:
PR for related tests:
https://github.com/MetaMask/metamask-extension/pull/23745
